services/slam/builtin/builtin_test.go

```diff
@@ -699,14 +698,14 @@ func TestCartographerDataProcess(t *testing.T) {
 		}()
 
 		cancelCtx, cancelFunc := context.WithCancel(context.Background())
-		slamSvc.StartDataProcess(cancelCtx, cams, camStreams, nil)
-
-		time.Sleep(time.Millisecond * time.Duration(validDataRateMS*2))
+		c := make(chan int)
+		slamSvc.StartDataProcess(cancelCtx, cams, camStreams, c)
 
+		<-c
 		allObs := obs.All()
 		latestLoggedEntry := allObs[len(allObs)-1]
-		test.That(t, fmt.Sprint(latestLoggedEntry), test.ShouldContainSubstring, "bad_lidar")
 		cancelFunc()
+		test.That(t, fmt.Sprint(latestLoggedEntry), test.ShouldContainSubstring, "bad_lidar")
```

Comment:
I'm also moving this check below the call to cancelFunc() so that cancelFunc() will be called even if this check fails.